diff -ruN clean-kernel/Kbuild work-kernel-v2/Kbuild
--- a/kernel/Kbuild	2026-04-09 11:57:48.449755460 +0100
+++ b/kernel/Kbuild	2026-04-10 01:45:24.335592689 +0100
@@ -5,9 +5,11 @@
 
 kernelsu-objs += hook/lsm_hook.o
 kernelsu-objs += hook/setuid_hook.o
+ifndef CONFIG_KSU_SUSFS
 kernelsu-objs += hook/syscall_event_bridge.o
 kernelsu-objs += hook/syscall_hook_manager.o
 kernelsu-objs += hook/tp_marker.o
+endif
 ifeq ($(CONFIG_ARM64),y)
 kernelsu-objs += hook/arm64/patch_memory.o
 kernelsu-objs += hook/arm64/syscall_hook.o
diff -ruN clean-kernel/Kconfig work-kernel-v2/Kconfig
--- a/kernel/Kconfig	2026-04-09 11:57:48.449755460 +0100
+++ b/kernel/Kconfig	2026-04-10 01:46:11.950555860 +0100
@@ -35,4 +35,85 @@
 	  Escalation will always use the default full root profile, and
 	  non-root handling will follow the global umount policy only.
 
+menu "KernelSU - SUSFS"
+config KSU_SUSFS
+	bool "KernelSU addon - SUSFS"
+	depends on KSU
+	depends on THREAD_INFO_IN_TASK
+	default y
+	help
+	  Patch and Enable SUSFS to kernel with KernelSU.
+
+config KSU_SUSFS_SUS_PATH
+	bool "Enable to hide suspicious path (NOT recommended)"
+	depends on KSU_SUSFS
+	default y
+	help
+	  Allow hiding the user-defined path and all its sub-paths from various
+	  system calls. Use with caution as it may cause performance loss.
+
+config KSU_SUSFS_SUS_MOUNT
+	bool "Enable to hide suspicious mounts"
+	depends on KSU_SUSFS
+	default y
+	help
+	  Allow hiding the user-defined mount paths from
+	  /proc/self/[mounts|mountinfo|mountstat].
+
+config KSU_SUSFS_SUS_KSTAT
+	bool "Enable to spoof suspicious kstat"
+	depends on KSU_SUSFS
+	default y
+	help
+	  Allow spoofing the kstat of user-defined file/directory.
+
+config KSU_SUSFS_SPOOF_UNAME
+	bool "Enable to spoof uname"
+	depends on KSU_SUSFS
+	default y
+	help
+	  Allow spoofing the string returned by uname syscall to user-defined
+	  string.
+
+config KSU_SUSFS_ENABLE_LOG
+	bool "Enable logging susfs log to kernel"
+	depends on KSU_SUSFS
+	default y
+	help
+	  Allow logging susfs log to kernel, uncheck it to completely disable
+	  all susfs log.
+
+config KSU_SUSFS_HIDE_KSU_SUSFS_SYMBOLS
+	bool "Enable to automatically hide ksu and susfs symbols from /proc/kallsyms"
+	depends on KSU_SUSFS
+	default y
+	help
+	  Automatically hide ksu and susfs symbols from /proc/kallsyms.
+
+config KSU_SUSFS_SPOOF_CMDLINE_OR_BOOTCONFIG
+	bool "Enable to spoof /proc/bootconfig (gki) or /proc/cmdline (non-gki)"
+	depends on KSU_SUSFS
+	default y
+	help
+	  Spoof the output of /proc/bootconfig (gki) or /proc/cmdline (non-gki)
+	  with a user-defined file.
+
+config KSU_SUSFS_OPEN_REDIRECT
+	bool "Enable to redirect a path to be opened with another path (experimental)"
+	depends on KSU_SUSFS
+	default y
+	help
+	  Allow redirecting a target path to be opened with another user-defined
+	  path.
+
+config KSU_SUSFS_SUS_MAP
+	bool "Enable to hide some mmapped real file from different proc maps interfaces"
+	depends on KSU_SUSFS
+	default y
+	help
+	  Allow hiding mmapped real file from
+	  /proc/<pid>/[maps|smaps|smaps_rollup|map_files|mem|pagemap].
+
+endmenu
+
 endmenu
diff -ruN clean-kernel/Makefile work-kernel-v2/Makefile
--- a/kernel/Makefile	2026-04-09 11:57:48.449755460 +0100
+++ b/kernel/Makefile	2026-04-10 01:45:59.786820883 +0100
@@ -22,4 +22,13 @@
 check-format:
 	find . \( -name "*.c" -o -name "*.h" \) -print0 | xargs -0 clang-format --dry-run --Werror
 
+## For susfs stuff ##
+ifeq ($(shell test -e $(srctree)/fs/susfs.c; echo $$?),0)
+$(eval SUSFS_VERSION=$(shell cat $(srctree)/include/linux/susfs.h | grep -E '^#define SUSFS_VERSION' | cut -d' ' -f3 | sed 's/"//g'))
+$(info )
+$(info -- SUSFS_VERSION: $(SUSFS_VERSION))
+else
+$(info -- You have not integrated susfs in your kernel yet.)
+$(info -- Read: https://gitlab.com/simonpunk/susfs4ksu)
+endif
 # Keep a new line here!! Because someone may append config
\ No newline at end of file
diff -ruN clean-kernel/core/init.c work-kernel-v2/core/init.c
--- a/kernel/core/init.c	2026-04-09 11:57:48.449755460 +0100
+++ b/kernel/core/init.c	2026-04-10 02:22:32.378905721 +0100
@@ -12,7 +12,14 @@
 #include "klog.h" // IWYU pragma: keep
 #include "manager/manager_observer.h"
 #include "manager/throne_tracker.h"
+#ifdef CONFIG_KSU_SUSFS
+#include <linux/susfs.h>
+#include "hook/setuid_hook.h"
+#include "feature/sucompat.h"
+#endif
+#ifndef CONFIG_KSU_SUSFS
 #include "hook/syscall_hook_manager.h"
+#endif
 #include "runtime/ksud.h"
 #include "runtime/ksud_boot.h"
 #include "supercall/supercall.h"
@@ -129,12 +136,21 @@
 		ksu_allowlist_init();
 		ksu_load_allow_list();
 
+#ifndef CONFIG_KSU_SUSFS
 		ksu_syscall_hook_manager_init();
+#else
+		ksu_setuid_hook_init();
+		ksu_sucompat_init();
+#endif
 
 		ksu_throne_tracker_init();
 		ksu_observer_init();
 		ksu_file_wrapper_init();
 
+#ifdef CONFIG_KSU_SUSFS
+		susfs_init();
+#endif
+
 		ksu_boot_completed = true;
 		track_throne(false);
 
@@ -144,13 +160,24 @@
 		}
 
 	} else {
+#ifndef CONFIG_KSU_SUSFS
 		ksu_syscall_hook_manager_init();
+#else
+		ksu_setuid_hook_init();
+		ksu_sucompat_init();
+#endif
 
 		ksu_allowlist_init();
 
 		ksu_throne_tracker_init();
 
+#ifdef CONFIG_KSU_SUSFS
+		susfs_init();
+#endif
+
+#ifndef CONFIG_KSU_SUSFS
 		ksu_ksud_init();
+#endif
 
 		ksu_file_wrapper_init();
 	}
@@ -165,13 +192,20 @@
 
 void __exit kernelsu_exit(void)
 {
+#ifndef CONFIG_KSU_SUSFS
 	// Phase 1: Stop all hooks first to prevent new callbacks
 	ksu_syscall_hook_manager_exit();
+#else
+	ksu_setuid_hook_exit();
+	ksu_sucompat_exit();
+#endif
 
 	ksu_supercalls_exit();
 
+#ifndef CONFIG_KSU_SUSFS
 	if (!ksu_late_loaded)
 		ksu_ksud_exit();
+#endif
 
 	// Wait for any in-flight RCU readers (e.g. handler traversing allow_list)
 	synchronize_rcu();
diff -ruN clean-kernel/feature/kernel_umount.c work-kernel-v2/feature/kernel_umount.c
--- a/kernel/feature/kernel_umount.c	2026-04-09 11:57:48.449755460 +0100
+++ b/kernel/feature/kernel_umount.c	2026-04-10 01:58:59.557397556 +0100
@@ -18,7 +18,7 @@
 #include "runtime/ksud_boot.h"
 #include "ksu.h"
 
-static bool ksu_kernel_umount_enabled = true;
+bool ksu_kernel_umount_enabled = true;
 
 static int kernel_umount_feature_get(u64 *value)
 {
@@ -87,6 +87,7 @@
 		return 0;
 	}
 
+#ifndef CONFIG_KSU_SUSFS
     // There are 6 scenarios:
     // 1. Normal app: zygote -> appuid
     // 2. Isolated process forked from zygote: zygote -> isolated_process
@@ -111,6 +112,7 @@
 		pr_info("handle umount ignore non zygote child: %d\n", current->pid);
 		return 0;
 	}
+#endif
 	// umount the target mnt
 	pr_info("handle umount for uid: %d, pid: %d\n", new_uid, current->pid);
 
diff -ruN clean-kernel/feature/sucompat.c work-kernel-v2/feature/sucompat.c
--- a/kernel/feature/sucompat.c	2026-04-09 11:57:48.450082382 +0100
+++ b/kernel/feature/sucompat.c	2026-04-10 02:22:13.266285204 +0100
@@ -11,6 +11,12 @@
 #include <linux/version.h>
 #include <linux/sched/task_stack.h>
 #include <linux/ptrace.h>
+#ifdef CONFIG_KSU_SUSFS
+#include <linux/susfs_def.h>
+#include <linux/namei.h>
+#include "selinux/selinux.h"
+#include "objsec.h"
+#endif
 
 #include "policy/allowlist.h"
 #include "policy/feature.h"
@@ -71,6 +77,7 @@
 	return userspace_stack_buffer(ksud_path, sizeof(ksud_path));
 }
 
+#ifndef CONFIG_KSU_SUSFS
 int ksu_handle_faccessat(int *dfd, const char __user **filename_user,
 		int *mode, int *__unused_flags)
 {
@@ -168,6 +175,126 @@
 do_orig_execve:
 	return ksu_syscall_table[orig_nr](regs);
 }
+#else /* CONFIG_KSU_SUSFS */
+static const char sh_path[] = SH_PATH;
+static const char su_path[] = SU_PATH;
+static const char ksud_path[] = KSUD_PATH;
+
+extern bool ksu_kernel_umount_enabled;
+
+/*
+ * return 0 -> No further checks should be required afterwards
+ * return 1 -> Further checks should be continued afterwards
+ */
+int ksu_handle_execveat_init(struct filename *filename)
+{
+	if (!filename || !filename->name)
+		return 1;
+
+	if (current->pid != 1 && is_init(get_current_cred())) {
+		if (unlikely(strcmp(filename->name, KSUD_PATH) == 0)) {
+			pr_info("hook_manager: escape to root for init executing ksud: %d\n",
+				current->pid);
+			escape_to_root_for_init();
+		} else if (likely(strstr(filename->name, "/app_process") == NULL &&
+				strstr(filename->name, "/adbd") == NULL) &&
+				!susfs_is_current_proc_umounted()) {
+			pr_info("susfs: mark no sucompat checks for pid: '%d', exec: '%s'\n",
+				current->pid, filename->name);
+			susfs_set_current_proc_umounted();
+		}
+		return 0;
+	}
+	return 1;
+}
+
+int ksu_handle_execveat_sucompat(int *fd, struct filename **filename_ptr,
+				 void *__never_use_argv,
+				 void *__never_use_envp,
+				 int *__never_use_flags)
+{
+	struct filename *filename;
+
+	if (unlikely(!filename_ptr))
+		return 0;
+
+	filename = *filename_ptr;
+	if (IS_ERR(filename))
+		return 0;
+
+	if (!ksu_handle_execveat_init(filename))
+		return 0;
+
+	if (likely(memcmp(filename->name, su_path, sizeof(su_path))))
+		return 0;
+
+	pr_info("ksu_handle_execveat_sucompat: su found\n");
+	memcpy((void *)filename->name, ksud_path, sizeof(ksud_path));
+
+	escape_with_root_profile();
+
+	return 0;
+}
+
+int ksu_handle_execveat(int *fd, struct filename **filename_ptr, void *argv,
+			void *envp, int *flags)
+{
+	return ksu_handle_execveat_sucompat(fd, filename_ptr, argv, envp,
+					    flags);
+}
+
+int ksu_handle_faccessat(int *dfd, const char __user **filename_user,
+			 int *mode, int *__unused_flags)
+{
+	char path[sizeof(su_path) + 1] = { 0 };
+
+	strncpy_from_user_nofault(path, *filename_user, sizeof(path));
+
+	if (unlikely(!memcmp(path, su_path, sizeof(su_path)))) {
+		pr_info("ksu_handle_faccessat: su->sh!\n");
+		*filename_user = sh_user_path();
+	}
+
+	return 0;
+}
+
+int ksu_handle_stat(int *dfd, struct filename **filename, int *flags)
+{
+	if (unlikely(IS_ERR(*filename) || (*filename)->name == NULL))
+		return 0;
+
+	if (likely(memcmp((*filename)->name, su_path, sizeof(su_path))))
+		return 0;
+
+	pr_info("ksu_handle_stat: su->sh!\n");
+	memcpy((void *)((*filename)->name), sh_path, sizeof(sh_path));
+	return 0;
+}
+
+int ksu_handle_devpts(struct inode *inode)
+{
+	uid_t uid;
+	struct inode_security_struct *sec;
+
+	if (!current->mm)
+		return 0;
+
+	uid = current_uid().val;
+	if (uid % 100000 < 10000)
+		return 0;
+
+	if (!__ksu_is_allow_uid_for_current(uid))
+		return 0;
+
+	if (ksu_file_sid) {
+		sec = selinux_inode(inode);
+		if (sec)
+			sec->sid = ksu_file_sid;
+	}
+
+	return 0;
+}
+#endif /* CONFIG_KSU_SUSFS */
 
 // sucompat: permitted process can execute 'su' to gain root access.
 void __init ksu_sucompat_init()
diff -ruN clean-kernel/feature/sucompat.h work-kernel-v2/feature/sucompat.h
--- a/kernel/feature/sucompat.h	2026-04-09 11:57:48.450082382 +0100
+++ b/kernel/feature/sucompat.h	2026-04-10 01:58:52.777549107 +0100
@@ -2,16 +2,29 @@
 #define __KSU_H_SUCOMPAT
 #include <asm/ptrace.h>
 #include <linux/types.h>
+#include <linux/version.h>
 
 extern bool ksu_su_compat_enabled;
 
 void ksu_sucompat_init(void);
 void ksu_sucompat_exit(void);
 
-// Handler functions exported for hook_manager
 int ksu_handle_faccessat(int *dfd, const char __user **filename_user, int *mode,
                          int *__unused_flags);
+
+#ifdef CONFIG_KSU_SUSFS
+#include <linux/fs.h>
+int ksu_handle_stat(int *dfd, struct filename **filename, int *flags);
+int ksu_handle_execveat_init(struct filename *filename);
+int ksu_handle_execveat_sucompat(int *fd, struct filename **filename_ptr,
+                                 void *argv, void *envp, int *flags);
+int ksu_handle_execveat(int *fd, struct filename **filename_ptr, void *argv,
+                        void *envp, int *flags);
+int ksu_handle_devpts(struct inode *inode);
+#else
 int ksu_handle_stat(int *dfd, const char __user **filename_user, int *flags);
-long ksu_handle_execve_sucompat(const char __user **filename_user, int orig_nr, const struct pt_regs *regs);
+long ksu_handle_execve_sucompat(const char __user **filename_user, int orig_nr,
+                                const struct pt_regs *regs);
+#endif
 
 #endif
diff -ruN clean-kernel/hook/setuid_hook.c work-kernel-v2/hook/setuid_hook.c
--- a/kernel/hook/setuid_hook.c	2026-04-09 11:57:48.450220824 +0100
+++ b/kernel/hook/setuid_hook.c	2026-04-10 02:21:51.107559729 +0100
@@ -11,6 +11,9 @@
 #include <linux/types.h>
 #include <linux/uaccess.h>
 #include <linux/uidgid.h>
+#ifdef CONFIG_KSU_SUSFS
+#include <linux/susfs_def.h>
+#endif
 
 #include "policy/allowlist.h"
 #include "hook/setuid_hook.h"
@@ -18,9 +21,35 @@
 #include "manager/manager_identity.h"
 #include "infra/seccomp_cache.h"
 #include "supercall/supercall.h"
+#ifndef CONFIG_KSU_SUSFS
 #include "hook/tp_marker.h"
+#endif
 #include "feature/kernel_umount.h"
+#ifdef CONFIG_KSU_SUSFS
+#include "selinux/selinux.h"
 
+static inline bool is_zygote_isolated_service_uid(uid_t uid)
+{
+	uid %= KSU_PER_USER_RANGE;
+	return (uid >= 99000 && uid < 100000);
+}
+
+static inline bool is_zygote_normal_app_uid(uid_t uid)
+{
+	uid %= KSU_PER_USER_RANGE;
+	return (uid >= 10000 && uid <= 19999);
+}
+
+extern u32 susfs_zygote_sid;
+#ifdef CONFIG_KSU_SUSFS_SUS_PATH
+extern void susfs_run_sus_path_loop(uid_t uid);
+#endif
+#ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
+extern void susfs_reorder_mnt_id(void);
+#endif
+#endif
+
+#ifndef CONFIG_KSU_SUSFS
 int ksu_handle_setresuid(uid_t old_uid, uid_t new_uid)
 {
     // we rely on the fact that zygote always call setresuid(3) with same uids
@@ -55,6 +84,59 @@
 
     return 0;
 }
+#else
+int ksu_handle_setresuid(uid_t old_uid, uid_t new_uid)
+{
+    if (!susfs_zygote_sid || !susfs_is_sid_equal(current_cred(), susfs_zygote_sid))
+        return 0;
+
+#ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
+    if (is_zygote_isolated_service_uid(new_uid))
+        goto do_umount;
+#endif
+
+    if (likely(ksu_is_manager_appid_valid()) &&
+        unlikely(ksu_get_manager_appid() == new_uid % KSU_PER_USER_RANGE)) {
+        spin_lock_irq(&current->sighand->siglock);
+        ksu_seccomp_allow_cache(current->seccomp.filter, __NR_reboot);
+        spin_unlock_irq(&current->sighand->siglock);
+
+        pr_info("install fd for manager: %d\n", new_uid);
+        ksu_install_fd();
+        return 0;
+    }
+
+    if (likely(is_zygote_normal_app_uid(new_uid) &&
+               ksu_uid_should_umount(new_uid)))
+        goto do_umount;
+
+    if (ksu_is_allow_uid_for_current(new_uid)) {
+        if (current->seccomp.mode == SECCOMP_MODE_FILTER &&
+            current->seccomp.filter) {
+            spin_lock_irq(&current->sighand->siglock);
+            ksu_seccomp_allow_cache(current->seccomp.filter, __NR_reboot);
+            spin_unlock_irq(&current->sighand->siglock);
+        }
+    }
+
+    return 0;
+
+do_umount:
+    ksu_handle_umount(old_uid, new_uid);
+
+#ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
+    susfs_reorder_mnt_id();
+#endif
+
+#ifdef CONFIG_KSU_SUSFS_SUS_PATH
+    susfs_run_sus_path_loop(new_uid);
+#endif
+
+    susfs_set_current_proc_umounted();
+
+    return 0;
+}
+#endif
 
 void __init ksu_setuid_hook_init(void)
 {
diff -ruN clean-kernel/policy/app_profile.c work-kernel-v2/policy/app_profile.c
--- a/kernel/policy/app_profile.c	2026-04-09 11:57:48.451239227 +0100
+++ b/kernel/policy/app_profile.c	2026-04-10 02:02:31.452661067 +0100
@@ -14,7 +14,9 @@
 #include "klog.h" // IWYU pragma: keep
 #include "selinux/selinux.h"
 #include "infra/su_mount_ns.h"
+#ifndef CONFIG_KSU_SUSFS
 #include "hook/tp_marker.h"
+#endif
 
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 7, 0)
 static struct group_info root_groups = { .usage = REFCOUNT_INIT(2) };
@@ -107,8 +109,10 @@
 {
     int ret = 0;
     struct cred *cred;
+#ifndef CONFIG_KSU_SUSFS
     struct task_struct *p = current;
     struct task_struct *t;
+#endif
     struct root_profile profile;
     struct user_struct *new_user;
 
@@ -185,9 +189,11 @@
 
     disable_seccomp();
 
+#ifndef CONFIG_KSU_SUSFS
     for_each_thread (p, t) {
         ksu_set_task_tracepoint_flag(t);
     }
+#endif
 
     setup_mount_ns(profile.namespaces);
     return 0;
diff -ruN clean-kernel/runtime/ksud.h work-kernel-v2/runtime/ksud.h
--- a/kernel/runtime/ksud.h	2026-04-09 11:57:48.451373296 +0100
+++ b/kernel/runtime/ksud.h	2026-04-10 02:04:45.505730821 +0100
@@ -11,4 +11,12 @@
 void ksu_execve_hook_ksud(const struct pt_regs *regs);
 void ksu_stop_input_hook_runtime(void);
 
+#ifdef CONFIG_KSU_SUSFS
+extern bool ksu_execveat_hook;
+extern bool ksu_read_hook;
+extern bool ksu_input_hook;
+void ksu_handle_sys_read(unsigned int fd);
+void ksu_handle_vfs_fstat(int fd, loff_t *kstat_size_ptr);
+#endif
+
 #endif
diff -ruN clean-kernel/runtime/ksud_integration.c work-kernel-v2/runtime/ksud_integration.c
--- a/kernel/runtime/ksud_integration.c	2026-04-09 11:57:48.451373296 +0100
+++ b/kernel/runtime/ksud_integration.c	2026-04-10 02:05:20.444976487 +0100
@@ -49,7 +49,13 @@
 static void stop_init_rc_hook();
 static void stop_execve_hook();
 
+#ifndef CONFIG_KSU_SUSFS
 static struct work_struct stop_input_hook_work;
+#else
+bool ksu_execveat_hook __read_mostly = true;
+bool ksu_read_hook __read_mostly = true;
+bool ksu_input_hook __read_mostly = true;
+#endif
 
 #define MAX_ARG_STRINGS 0x7FFFFFFF
 struct user_arg_ptr {
@@ -316,8 +322,16 @@
     file->f_op = &fops_proxy;
 }
 
+#ifdef CONFIG_KSU_SUSFS
+void ksu_handle_sys_read(unsigned int fd)
+#else
 static void ksu_handle_sys_read(unsigned int fd, char __user **buf_ptr, size_t *count_ptr)
+#endif
 {
+#ifdef CONFIG_KSU_SUSFS
+    if (!ksu_read_hook)
+        return;
+#endif
     struct file *file = fget(fd);
     if (!file) {
         return;
@@ -335,6 +349,11 @@
 
 int ksu_handle_input_handle_event(unsigned int *type, unsigned int *code, int *value)
 {
+#ifdef CONFIG_KSU_SUSFS
+    if (!ksu_input_hook)
+        return 0;
+#endif
+
     if (*type == EV_KEY && *code == KEY_VOLUMEDOWN) {
         int val = *value;
         pr_info("KEY_VOLUMEDOWN val: %d\n", val);
@@ -378,6 +397,10 @@
 
 void ksu_execve_hook_ksud(const struct pt_regs *regs)
 {
+#ifdef CONFIG_KSU_SUSFS
+    if (!ksu_execveat_hook)
+        return;
+#endif
     const char __user **filename_user = (const char **)&PT_REGS_PARM1(regs);
     const char __user *const __user *__argv = (const char __user *const __user *)PT_REGS_PARM2(regs);
     struct user_arg_ptr argv = { .ptr.native = __argv };
@@ -406,10 +429,14 @@
 static long ksu_sys_read(const struct pt_regs *regs)
 {
     unsigned int fd = PT_REGS_PARM1(regs);
+#ifndef CONFIG_KSU_SUSFS
     char __user **buf_ptr = (char __user **)&PT_REGS_PARM2(regs);
     size_t *count_ptr = (size_t *)&PT_REGS_PARM3(regs);
 
     ksu_handle_sys_read(fd, buf_ptr, count_ptr);
+#else
+    ksu_handle_sys_read(fd);
+#endif
     return orig_sys_read(regs);
 }
 
@@ -451,6 +478,25 @@
     return ret;
 }
 
+#ifdef CONFIG_KSU_SUSFS
+void ksu_handle_vfs_fstat(int fd, loff_t *kstat_size_ptr)
+{
+    loff_t new_size = *kstat_size_ptr + ksu_rc_len;
+    struct file *file = fget(fd);
+
+    if (!file)
+        return;
+
+    if (is_init_rc(file)) {
+        pr_info("stat init.rc");
+        pr_info("adding ksu_rc_len: %lld -> %lld", *kstat_size_ptr, new_size);
+        *kstat_size_ptr = new_size;
+    }
+    fput(file);
+}
+#endif
+
+#ifndef CONFIG_KSU_SUSFS
 static int input_handle_event_handler_pre(struct kprobe *p, struct pt_regs *regs)
 {
     unsigned int *type = (unsigned int *)&PT_REGS_PARM2(regs);
@@ -468,6 +514,7 @@
 {
     unregister_kprobe(&input_event_kp);
 }
+#endif
 
 static void stop_init_rc_hook()
 {
@@ -483,28 +530,37 @@
         return;
     }
     input_hook_stopped = true;
+#ifndef CONFIG_KSU_SUSFS
     bool ret = schedule_work(&stop_input_hook_work);
     pr_info("unregister input kprobe: %d!\n", ret);
+#else
+    ksu_input_hook = false;
+    pr_info("stop input_hook\n");
+#endif
 }
 
 // ksud: module support
 void __init ksu_ksud_init()
 {
-    int ret;
-
     ksu_syscall_table_hook(__NR_read, ksu_sys_read, &orig_sys_read);
     ksu_syscall_table_hook(__NR_fstat, ksu_sys_fstat, &orig_sys_fstat);
 
-    ret = register_kprobe(&input_event_kp);
-    pr_info("ksud: input_event_kp: %d\n", ret);
+#ifndef CONFIG_KSU_SUSFS
+    {
+        int ret = register_kprobe(&input_event_kp);
+        pr_info("ksud: input_event_kp: %d\n", ret);
+    }
 
     INIT_WORK(&stop_input_hook_work, do_stop_input_hook);
+#endif
 }
 
 void __exit ksu_ksud_exit()
 {
+#ifndef CONFIG_KSU_SUSFS
     // TODO:
     // this should be done before unregister vfs_read_kp
     // stop_init_rc_hook();
     unregister_kprobe(&input_event_kp);
+#endif
 }
diff -ruN clean-kernel/selinux/rules.c work-kernel-v2/selinux/rules.c
--- a/kernel/selinux/rules.c	2026-04-09 11:57:48.451373296 +0100
+++ b/kernel/selinux/rules.c	2026-04-10 01:50:24.392913022 +0100
@@ -119,6 +119,15 @@
     ksu_allow(db, "system_server", KERNEL_SU_DOMAIN, "process", "getpgid");
     ksu_allow(db, "system_server", KERNEL_SU_DOMAIN, "process", "sigkill");
 
+#ifdef CONFIG_KSU_SUSFS
+    // Allow umount in zygote process without installing zygisk
+    //ksu_allow(db, "zygote", "labeledfs", "filesystem", "unmount");
+    susfs_set_priv_app_sid();
+    susfs_set_init_sid();
+    susfs_set_ksu_sid();
+    susfs_set_zygote_sid();
+#endif
+
     rcu_assign_pointer(selinux_state.policy, pol);
     synchronize_rcu();
     ksu_destroy_sepolicy(old_pol);
diff -ruN clean-kernel/selinux/selinux.c work-kernel-v2/selinux/selinux.c
--- a/kernel/selinux/selinux.c	2026-04-09 11:57:48.451550590 +0100
+++ b/kernel/selinux/selinux.c	2026-04-10 01:49:07.846624059 +0100
@@ -207,3 +207,99 @@
 {
     return is_sid_match(cred, cached_init_sid, INIT_CONTEXT);
 }
+
+#ifdef CONFIG_KSU_SUSFS
+#define KERNEL_INIT_DOMAIN "u:r:init:s0"
+#define KERNEL_ZYGOTE_DOMAIN "u:r:zygote:s0"
+#define KERNEL_PRIV_APP_DOMAIN "u:r:priv_app:s0:c512,c768"
+
+u32 susfs_ksu_sid = 0;
+u32 susfs_init_sid = 0;
+u32 susfs_zygote_sid = 0;
+u32 susfs_priv_app_sid = 0;
+
+static inline void susfs_set_sid(const char *secctx_name, u32 *out_sid)
+{
+	int err;
+
+	if (!secctx_name || !out_sid) {
+		pr_err("secctx_name || out_sid is NULL\n");
+		return;
+	}
+
+	err = security_secctx_to_secid(secctx_name, strlen(secctx_name),
+				       out_sid);
+	if (err) {
+		pr_err("failed setting sid for '%s', err: %d\n", secctx_name, err);
+		return;
+	}
+	pr_info("sid '%u' is set for secctx_name '%s'\n", *out_sid, secctx_name);
+}
+
+bool susfs_is_sid_equal(const struct cred *cred, u32 sid2) {
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 18, 0)
+	const struct task_security_struct *tsec = selinux_cred(cred);
+#else
+	const struct cred_security_struct *tsec = selinux_cred(cred);
+#endif
+
+	if (!tsec) {
+		return false;
+	}
+	return tsec->sid == sid2;
+}
+
+u32 susfs_get_sid_from_name(const char *secctx_name)
+{
+	u32 out_sid = 0;
+	int err;
+
+	if (!secctx_name) {
+		pr_err("secctx_name is NULL\n");
+		return 0;
+	}
+	err = security_secctx_to_secid(secctx_name, strlen(secctx_name),
+				       &out_sid);
+	if (err) {
+		pr_err("failed getting sid from secctx_name: %s, err: %d\n", secctx_name, err);
+		return 0;
+	}
+	return out_sid;
+}
+
+u32 susfs_get_current_sid(void) {
+	return current_sid();
+}
+
+void susfs_set_zygote_sid(void)
+{
+	susfs_set_sid(KERNEL_ZYGOTE_DOMAIN, &susfs_zygote_sid);
+}
+
+bool susfs_is_current_zygote_domain(void) {
+	return unlikely(current_sid() == susfs_zygote_sid);
+}
+
+void susfs_set_ksu_sid(void)
+{
+	susfs_set_sid(KERNEL_SU_CONTEXT, &susfs_ksu_sid);
+}
+
+bool susfs_is_current_ksu_domain(void) {
+	return unlikely(current_sid() == susfs_ksu_sid);
+}
+
+void susfs_set_init_sid(void)
+{
+	susfs_set_sid(KERNEL_INIT_DOMAIN, &susfs_init_sid);
+}
+
+bool susfs_is_current_init_domain(void) {
+	return unlikely(current_sid() == susfs_init_sid);
+}
+
+void susfs_set_priv_app_sid(void)
+{
+	susfs_set_sid(KERNEL_PRIV_APP_DOMAIN, &susfs_priv_app_sid);
+}
+#endif // #ifdef CONFIG_KSU_SUSFS
diff -ruN clean-kernel/selinux/selinux.h work-kernel-v2/selinux/selinux.h
--- a/kernel/selinux/selinux.h	2026-04-09 11:57:48.451550590 +0100
+++ b/kernel/selinux/selinux.h	2026-04-10 01:47:43.900500506 +0100
@@ -37,4 +37,17 @@
 
 extern u32 ksu_file_sid;
 
+#ifdef CONFIG_KSU_SUSFS
+bool susfs_is_sid_equal(const struct cred *cred, u32 sid2);
+u32 susfs_get_sid_from_name(const char *secctx_name);
+u32 susfs_get_current_sid(void);
+void susfs_set_zygote_sid(void);
+bool susfs_is_current_zygote_domain(void);
+void susfs_set_ksu_sid(void);
+bool susfs_is_current_ksu_domain(void);
+void susfs_set_init_sid(void);
+bool susfs_is_current_init_domain(void);
+void susfs_set_priv_app_sid(void);
+#endif
+
 #endif
diff -ruN clean-kernel/supercall/dispatch.c work-kernel-v2/supercall/dispatch.c
--- a/kernel/supercall/dispatch.c	2026-04-09 11:57:48.451550590 +0100
+++ b/kernel/supercall/dispatch.c	2026-04-10 02:08:19.989100156 +0100
@@ -4,6 +4,9 @@
 #include <linux/string.h>
 #include <linux/uaccess.h>
 #include <linux/version.h>
+#ifdef CONFIG_KSU_SUSFS
+#include <linux/susfs.h>
+#endif
 
 #include "uapi/supercall.h"
 #include "supercall/internal.h"
@@ -17,7 +20,9 @@
 #include "manager/manager_identity.h"
 #include "selinux/selinux.h"
 #include "infra/file_wrapper.h"
+#ifndef CONFIG_KSU_SUSFS
 #include "hook/tp_marker.h"
+#endif
 #include "policy/app_profile.h"
 #include "supercall/supercall.h"
 
@@ -96,6 +101,9 @@
             } else {
                 pr_info("boot_complete triggered\n");
                 on_boot_completed();
+#ifdef CONFIG_KSU_SUSFS
+                susfs_start_sdcard_monitor_fn();
+#endif
             }
         }
         break;
@@ -335,7 +343,9 @@
     ret = ksu_set_app_profile(&cmd.profile);
     if (!ret) {
         ksu_persistent_allow_list();
+#ifndef CONFIG_KSU_SUSFS
         ksu_mark_running_process();
+#endif
     }
     return ret;
 }
@@ -413,6 +423,7 @@
 
     switch (cmd.operation) {
     case KSU_MARK_GET: {
+#ifndef CONFIG_KSU_SUSFS
         // Get task mark status
         ret = ksu_get_task_mark(cmd.pid);
         if (ret < 0) {
@@ -421,8 +432,19 @@
         }
         cmd.result = (u32)ret;
         break;
+#else
+        if (susfs_is_current_proc_umounted()) {
+            ret = 0; // SYSCALL_TRACEPOINT is NOT flagged
+        } else {
+            ret = 1; // SYSCALL_TRACEPOINT is flagged
+        }
+        pr_info("manage_mark: ret for pid %d: %d\n", cmd.pid, ret);
+        cmd.result = (u32)ret;
+        break;
+#endif
     }
     case KSU_MARK_MARK: {
+#ifndef CONFIG_KSU_SUSFS
         if (cmd.pid == 0) {
             ksu_mark_all_process();
         } else {
@@ -432,9 +454,15 @@
                 return ret;
             }
         }
+#else
+        if (cmd.pid != 0) {
+            return ret;
+        }
+#endif
         break;
     }
     case KSU_MARK_UNMARK: {
+#ifndef CONFIG_KSU_SUSFS
         if (cmd.pid == 0) {
             ksu_unmark_all_process();
         } else {
@@ -444,11 +472,20 @@
                 return ret;
             }
         }
+#else
+        if (cmd.pid != 0) {
+            return ret;
+        }
+#endif
         break;
     }
     case KSU_MARK_REFRESH: {
+#ifndef CONFIG_KSU_SUSFS
         ksu_mark_running_process();
         pr_info("manage_mark: refreshed running processes\n");
+#else
+        pr_info("susfs: cmd: KSU_MARK_REFRESH: do nothing\n");
+#endif
         break;
     }
     default: {
diff -ruN clean-kernel/supercall/supercall.c work-kernel-v2/supercall/supercall.c
--- a/kernel/supercall/supercall.c	2026-04-09 11:57:48.451895147 +0100
+++ b/kernel/supercall/supercall.c	2026-04-10 02:22:15.122386813 +0100
@@ -11,6 +11,10 @@
 #include <linux/uaccess.h>
 #include <linux/version.h>
 #include <linux/utsname.h> // utsname() and uts_sem
+#ifdef CONFIG_KSU_SUSFS
+#include <linux/susfs.h>
+#include <linux/namei.h>
+#endif
 
 #include "uapi/supercall.h"
 #include "supercall/internal.h"
@@ -86,6 +90,7 @@
     kfree(tw);
 }
 
+#ifndef CONFIG_KSU_SUSFS
 static int reboot_handler_pre(struct kprobe *p, struct pt_regs *regs)
 {
     struct pt_regs *real_regs = PT_REAL_REGS(regs);
@@ -232,25 +237,244 @@
     .symbol_name = REBOOT_SYMBOL,
     .pre_handler = reboot_handler_pre,
 };
+#else
+int ksu_handle_sys_reboot(int magic1, int magic2, unsigned int cmd, void __user **arg)
+{
+    if (magic1 != KSU_INSTALL_MAGIC1)
+        return -EINVAL;
+
+    if (!arg)
+        return -EINVAL;
+
+    if (magic2 == SUSFS_MAGIC && current_uid().val == 0) {
+#ifdef CONFIG_KSU_SUSFS_SUS_PATH
+        if (cmd == CMD_SUSFS_ADD_SUS_PATH) {
+            susfs_add_sus_path(arg);
+            return 0;
+        }
+        if (cmd == CMD_SUSFS_ADD_SUS_PATH_LOOP) {
+            susfs_add_sus_path_loop(arg);
+            return 0;
+        }
+        if (cmd == CMD_SUSFS_SET_ANDROID_DATA_ROOT_PATH) {
+            susfs_set_i_state_on_external_dir(arg);
+            return 0;
+        }
+        if (cmd == CMD_SUSFS_SET_SDCARD_ROOT_PATH) {
+            susfs_set_i_state_on_external_dir(arg);
+            return 0;
+        }
+#endif
+#ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
+        if (cmd == CMD_SUSFS_HIDE_SUS_MNTS_FOR_NON_SU_PROCS) {
+            susfs_set_hide_sus_mnts_for_non_su_procs(arg);
+            return 0;
+        }
+#endif
+#ifdef CONFIG_KSU_SUSFS_SUS_KSTAT
+        if (cmd == CMD_SUSFS_ADD_SUS_KSTAT) {
+            susfs_add_sus_kstat(arg);
+            return 0;
+        }
+        if (cmd == CMD_SUSFS_UPDATE_SUS_KSTAT) {
+            susfs_update_sus_kstat(arg);
+            return 0;
+        }
+        if (cmd == CMD_SUSFS_ADD_SUS_KSTAT_STATICALLY) {
+            susfs_add_sus_kstat(arg);
+            return 0;
+        }
+#endif
+#ifdef CONFIG_KSU_SUSFS_SPOOF_UNAME
+        if (cmd == CMD_SUSFS_SET_UNAME) {
+            susfs_set_uname(arg);
+            return 0;
+        }
+#endif
+#ifdef CONFIG_KSU_SUSFS_ENABLE_LOG
+        if (cmd == CMD_SUSFS_ENABLE_LOG) {
+            susfs_enable_log(arg);
+            return 0;
+        }
+#endif
+#ifdef CONFIG_KSU_SUSFS_SPOOF_CMDLINE_OR_BOOTCONFIG
+        if (cmd == CMD_SUSFS_SET_CMDLINE_OR_BOOTCONFIG) {
+            susfs_set_cmdline_or_bootconfig(arg);
+            return 0;
+        }
+#endif
+#ifdef CONFIG_KSU_SUSFS_OPEN_REDIRECT
+        if (cmd == CMD_SUSFS_ADD_OPEN_REDIRECT) {
+            susfs_add_open_redirect(arg);
+            return 0;
+        }
+#endif
+#ifdef CONFIG_KSU_SUSFS_SUS_MAP
+        if (cmd == CMD_SUSFS_ADD_SUS_MAP) {
+            susfs_add_sus_map(arg);
+            return 0;
+        }
+#endif
+        if (cmd == CMD_SUSFS_ENABLE_AVC_LOG_SPOOFING) {
+            susfs_set_avc_log_spoofing(arg);
+            return 0;
+        }
+        if (cmd == CMD_SUSFS_SHOW_ENABLED_FEATURES) {
+            susfs_get_enabled_features(arg);
+            return 0;
+        }
+        if (cmd == CMD_SUSFS_SHOW_VARIANT) {
+            susfs_show_variant(arg);
+            return 0;
+        }
+        if (cmd == CMD_SUSFS_SHOW_VERSION) {
+            susfs_show_version(arg);
+            return 0;
+        }
+        return 0;
+    }
+
+    if (magic2 == KSU_INSTALL_MAGIC2) {
+        struct ksu_install_fd_tw *tw;
+
+        tw = kzalloc(sizeof(*tw), GFP_ATOMIC);
+        if (!tw)
+            return 0;
+
+        tw->outp = (int __user *)(*arg);
+        tw->cb.func = ksu_install_fd_tw_func;
+
+        if (task_work_add(current, &tw->cb, TWA_RESUME)) {
+            kfree(tw);
+            pr_warn("install fd add task_work failed\n");
+        }
+    }
+
+    if (magic2 == CHANGE_MANAGER_UID) {
+        if (current_uid().val != 0)
+            return 0;
+
+        pr_info("sys_reboot: ksu_set_manager_appid to: %d\n", cmd);
+        ksu_set_manager_appid(cmd);
+
+        if (cmd == ksu_get_manager_appid()) {
+            unsigned long reply = (unsigned long)(*arg);
+            if (copy_to_user((void __user *)(*arg), &reply, sizeof(reply)))
+                pr_info("sys_reboot: reply fail\n");
+        }
+        return 0;
+    }
+
+    if (magic2 == GET_SULOG_DUMP_V2) {
+        if (current_uid().val != 0)
+            return 0;
+
+        int ret = send_sulog_dump((void __user *)(*arg));
+        if (ret)
+            return 0;
+
+        unsigned long reply = (unsigned long)(*arg);
+        if (copy_to_user((void __user *)(*arg), &reply, sizeof(reply)))
+            return 0;
+    }
+
+    if (magic2 == CHANGE_KSUVER) {
+        if (current_uid().val != 0)
+            return 0;
+
+        pr_info("sys_reboot: ksu_change_ksuver to: %d\n", cmd);
+        ksuver_override = cmd;
+
+        unsigned long reply = (unsigned long)(*arg);
+        if (copy_to_user((void __user *)(*arg), &reply, sizeof(reply)))
+            return 0;
+    }
+
+    if (magic2 == CHANGE_SPOOF_UNAME) {
+        if (current_uid().val != 0)
+            return 0;
+
+        char release_buf[65];
+        char version_buf[65];
+        static char original_release_buf[65] = {0};
+        static char original_version_buf[65] = {0};
+
+        void __user **ppptr = (void __user **)(*arg);
+
+        uint64_t u_pptr = 0;
+        uint64_t u_ptr = 0;
+
+        pr_info("sys_reboot: ppptr: 0x%lx \n", (uintptr_t)ppptr);
+
+        if (copy_from_user(&u_pptr, ppptr, sizeof(u_pptr)))
+            return 0;
+
+        pr_info("sys_reboot: u_pptr: 0x%lx \n", (uintptr_t)u_pptr);
+
+        if (copy_from_user(&u_ptr, (void __user *)u_pptr, sizeof(u_ptr)))
+            return 0;
+
+        pr_info("sys_reboot: u_ptr: 0x%lx \n", (uintptr_t)u_ptr);
+
+        if (strncpy_from_user(release_buf, (char __user *)u_ptr, sizeof(release_buf)) < 0)
+            return 0;
+        release_buf[sizeof(release_buf) - 1] = '\0';
+
+        if (strncpy_from_user(version_buf, (char __user *)(u_ptr + strlen(release_buf) + 1), sizeof(version_buf)) < 0)
+            return 0;
+        version_buf[sizeof(version_buf) - 1] = '\0';
+
+        if (original_release_buf[0] == '\0') {
+            struct new_utsname *u_curr = utsname();
+            strscpy(original_release_buf, u_curr->release, sizeof(original_release_buf));
+            strscpy(original_version_buf, u_curr->version, sizeof(original_version_buf));
+            pr_info("sys_reboot: original uname saved: %s %s\n", original_release_buf, original_version_buf);
+        }
+
+        if (!strcmp(release_buf, "default") || !strcmp(version_buf, "default") ) {
+            memcpy(release_buf, original_release_buf, sizeof(release_buf));
+            memcpy(version_buf, original_version_buf, sizeof(version_buf));
+        }
+
+        pr_info("sys_reboot: spoofing kernel to: %s - %s\n", release_buf, version_buf);
+
+        struct new_utsname *u = utsname();
+
+        down_write(&uts_sem);
+        strscpy(u->release, release_buf, sizeof(u->release));
+        strscpy(u->version, version_buf, sizeof(u->version));
+        up_write(&uts_sem);
+
+        unsigned long reply = (unsigned long)(*arg);
+        if (copy_to_user((void __user *)(*arg), &reply, sizeof(reply)))
+            return 0;
+    }
+
+    return 0;
+}
+#endif
 
 void __init ksu_supercalls_init(void)
 {
-    int rc;
-
     ksu_supercall_dump_commands();
 
-    rc = register_kprobe(&reboot_kp);
-    if (rc) {
-        pr_err("reboot kprobe failed: %d\n", rc);
-    } else {
-        pr_info("reboot kprobe registered successfully\n");
+#ifndef CONFIG_KSU_SUSFS
+    {
+        int rc = register_kprobe(&reboot_kp);
+        if (rc)
+            pr_err("reboot kprobe failed: %d\n", rc);
+        else
+            pr_info("reboot kprobe registered successfully\n");
     }
+#endif
 
     sulog_init_heap(); // grab heap memory
 }
 
 void __exit ksu_supercalls_exit(void)
 {
+#ifndef CONFIG_KSU_SUSFS
     unregister_kprobe(&reboot_kp);
+#endif
     ksu_supercall_cleanup_state();
 }
